ngraph_bridge/ngraph_data_cache.h

```diff
@@ -0,0 +1,180 @@
+/*******************************************************************************
+ * Copyright 2019 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+
+#ifndef NGRAPH_DATA_CACHE_H_
+#define NGRAPH_DATA_CACHE_H_
+#pragma once
+
+#include <deque>
+#include <mutex>
+#include <ostream>
+#include <vector>
+#include "absl/synchronization/mutex.h"
+
+#include "tensorflow/core/framework/tensor_shape.h"
+#include "tensorflow/core/graph/graph.h"
+
+#include "logging/ngraph_log.h"
+#include "ngraph/ngraph.hpp"
+
+#include "ngraph_bridge/ngraph_freshness_tracker.h"
+#include "ngraph_bridge/ngraph_pipelined_tensors.h"
+
+namespace tensorflow {
+
+namespace ngraph_bridge {
+
+// Forward declaration for friend class
+namespace testing {
+class NGraphDataCacheTest_SameKeyMultiThread_Test;
+class NGraphDataCacheTest_RemoveItemTest_Test;
+}
+
+template <typename KeyType, typename ValueType>
+class NgraphDataCache {
+ public:
+  explicit NgraphDataCache(int depth);
+  ~NgraphDataCache();
+
+  // This method performs lookup in the cache for requested key, if not found
+  // it will create item, put it in the cache and returns item and status
+  std::pair<Status, ValueType> LookUpOrCreate(
+      KeyType key,
+      std::function<std::pair<Status, ValueType>()> callback_create_item,
+      std::function<void(ValueType)> callback_destroy_item);
+
+  // Overload for above function, which doesn't take callback for destroy item
+  std::pair<Status, ValueType> LookUpOrCreate(
+      KeyType key,
+      std::function<std::pair<Status, ValueType>()> callback_create_item);
+  void RemoveItem(KeyType key);
+  void RemoveItem(KeyType key,
+                  std::function<void(ValueType)> callback_destroy_item);
+  void RemoveAll(std::function<void(ValueType)> callback_destroy_item);
+
+ private:
+  std::unordered_map<KeyType, ValueType> m_ng_items_map;
+  std::deque<KeyType> m_lru;
```

Comment:
Does it make sense to have an assert soemwhere( in LookUpOrCreate/RemoveItmes) to check that m_ng_items_map.size()== m_lru.size().  they should always be the same. and we might find error cases

Reply:
@sayantan-nervana @avijit-nervana